src/ray/gcs/gcs_server/gcs_job_manager.cc

```diff
@@ -114,5 +159,117 @@ void GcsJobManager::HandleReportJobError(const rpc::ReportJobErrorRequest &reque
   GCS_RPC_SEND_REPLY(send_reply_callback, reply, Status::OK());
 }
 
+void GcsJobManager::HandleSubmitJob(const rpc::SubmitJobRequest &request,
+                                    rpc::SubmitJobReply *reply,
+                                    rpc::SendReplyCallback send_reply_callback) {
+  auto on_done = [reply, send_reply_callback](const Status &status) {
+    GCS_RPC_SEND_REPLY(send_reply_callback, reply, status);
+  };
+  auto status = SubmitJob(request, on_done);
+  if (!status.ok()) {
+    GCS_RPC_SEND_REPLY(send_reply_callback, reply, status);
+  }
+}
+
+Status GcsJobManager::SubmitJob(const ray::rpc::SubmitJobRequest &request,
+                                const ray::gcs::StatusCallback &callback) {
+  auto job_id = JobID::FromBinary(request.job_id());
+
+  RAY_LOG(INFO) << "Submitting job " << job_id;
+
+  auto it = jobs_.find(job_id);
+  if (it != jobs_.end()) {
+    RAY_LOG(ERROR) << "Failed to submit job " << job_id << ", job id conflicts.";
+    std::ostringstream ss;
+    ss << "Job id conflicts: " << job_id;
+    return Status::Invalid(ss.str());
+  }
+
+  // Fill in as much information as possible for the dashboard to
+  // show the job info in SUBMITTED state.
+  auto job_table_data = std::make_shared<rpc::JobTableData>();
+  job_table_data->set_state(rpc::JobTableData::SUBMITTED);
+  job_table_data->set_job_id(request.job_id());
+  job_table_data->set_language(request.language());
+  // Set the job payload (the json submitted from dashboard).
+  job_table_data->set_job_payload(request.job_payload());
+  // Mark the job is submitted, this field determines whether to
+  // initialize the job environement or not.
```

Comment:
"This flag is used to determine whether we need to initialize the job environment or not."

Reply:
Fixed.